crates/weaver_forge/src/jq.rs

```diff
@@ -0,0 +1,221 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Library to hide details of jaq from the rest of weaver.
+
+use std::collections::BTreeMap;
+
+use crate::error::Error;
+use jaq_core::{
+    load::{parse::Def, Arena, File, Loader},
+    Ctx, Native, RcIter,
+};
+use jaq_json::Val;
+
+type JqFileType = ();
+
+fn semconv_prelude() -> impl Iterator<Item = Def<&'static str>> {
+    jaq_core::load::parse(crate::SEMCONV_JQ, |p| p.defs())
+        .expect("BAD WEAVER BUILD - default JQ library failed to compile")
+        .into_iter()
+}
+
+fn prepare_jq_context(params: &BTreeMap<String, serde_json::Value>) -> (Vec<String>, Vec<Val>) {
+    let (jq_vars, jq_ctx): (Vec<String>, Vec<Val>) = params
+        .iter()
+        .map(|(k, v)| (format!("${k}"), Val::from(v.clone())))
+        .unzip();
+    (jq_vars, jq_ctx)
+}
+
+/// This is our single entry point for calling into the jaq library to run jq filters.
+pub fn execute_jq(
+    // The JSON input to JQ.
+    input: &serde_json::Value,
+    // The JQ filter to compile.
+    filter_expr: &str,
+    // Note: This will be exposed with `${key}` as the variable name.
+    params: &BTreeMap<String, serde_json::Value>,
+) -> Result<serde_json::Value, Error> {
+    let loader = Loader::new(
+        // ToDo: Allow custom preludes?
+        jaq_std::defs()
+            .chain(jaq_json::defs())
+            .chain(semconv_prelude()), // [],
+    );
+    let arena = Arena::default();
+    let program: File<&str, JqFileType> = File {
+        code: filter_expr,
+        path: (), // ToDo - give this the weaver-config location.
+    };
+
+    // parse the filter
+    let modules = loader
+        .load(&arena, program)
+        .map_err(load_errors)
+        .map_err(|e| Error::FilterError {
+            filter: filter_expr.to_owned(),
+            error: e,
+        })?;
+
+    let (names, values) = prepare_jq_context(params);
+    let funs = jaq_std::funs().chain(jaq_json::funs());
```

Comment:
nit: I’m not entirely sure, but you could try something like this to make the intent more explicit.

```
fn adjust_lifetime<'a, T: 'a>(iter: impl Iterator<Item = T>) -> impl Iterator<Item = T> + 'a {
    iter
}
 
let funs = adjust_lifetime(jaq_std::funs().chain(jaq_json::funs()));
...
  .with_funs(funs)
  .compile(modules)
  ...
```

Reply:
unfortunately, that function doesn't work (`iter` may not live for lifetime of `'a`).  What we need is to trick the compiler to infer a narrower lifetime when seeing both `'a` and `'static` to return `'a`, and we can only do that when they are in contravariant (or return-value) position.  I haven't figured out how to do that yet, but I can look that up for future cleanup. I figure it's not important enough to block this PR.

Reply:
So i think I have my arrows backwards (using contravariant to mean covariant), but otherwise (while I didn't get a chance to fully read this), the answer probably lies here: https://doc.rust-lang.org/nomicon/subtyping.html

From what I can tell the variance rules are only used to change lifetimes within rust, the trick I'm using with lambda is the only way I could think of.